File: Week06/LeetCode-0347-1.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        """This is an O(nlogn) solution"""
        nums.sort()
        
        freqs = []
        num = nums[0]
        freq = 1
        for i in range(1, len(nums)):
            if nums[i] == num:
                freq += 1
            else:
                freqs.append([freq, num])
                num = nums[i]
                freq = 1
        freqs.append([freq, num])
        
        # Sort freqs array wrt the frequencies in descending order
        freqs.sort(reverse=True)
        
        # Take the first "k" numbers
        result = []
        for i in range(k):
            result.append(freqs[i][1])
        return result

File: Week06/LeetCode-0347-1.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        /// This is an O(nlogn) solution
        std::sort(nums.begin(), nums.end());

        vector<pair<int, int>> freqs;
        int num = nums[0];
        int freq = 1;
        for (int i=1; i<nums.size(); i++){
            if (nums[i] == num) freq++;
            else {
                freqs.push_back({freq, num});
                num = nums[i];
                freq = 1;
            } //end-else
        } //end-for
        freqs.push_back({freq, num});
        
        // Sort freqs array wrt the frequencies in descending order
        std::sort(freqs.begin(), freqs.end(), std::greater());
                
        //Take the first "k" numbers
        vector<int> result;
        for (int i=0; i<k; i++)
            result.push_back(freqs[i].second);
        return result;
    } // end-topKFrequent
};